src/bun.js/webcore/encoding.zig

```diff
@@ -549,46 +550,96 @@ pub const TextEncoderStreamEncoder = struct {
 
                 break :prepend Prepend.replacement;
             }
-            break :prepend null;
+
+            break :prepend Prepend.empty;
         };
 
-        const length = bun.simdutf.length.utf8.from.utf16.le(remain);
+        bun.debugAssert(remain.len != 0);
+
+        const last_token = remain[remain.len - 1];
+        if (strings.u16IsLead(last_token)) {
+
+            // if last unit is high surrogate, exclude from simdutf conversion to avoid error
+            remain = remain[0 .. remain.len - 1];
+            this.pending_lead_surrogate = last_token;
+
+            if (remain.len == 0) {
+                if (prepend.len == 0) {
+                    return .undefined;
+                }
+
+                return ArrayBuffer.createBuffer(
+                    globalObject,
+                    prepend.bytes[0..prepend.len],
+                );
+            }
+        }
+
+        bun.debugAssert(remain.len != 0);
+
+        const max_length = ((remain.len + 1) * 3) + prepend.len;
 
-        var buf = std.ArrayList(u8).initCapacity(
-            bun.default_allocator,
-            length + @as(usize, if (prepend) |pre| pre.len else 0),
-        ) catch {
+        const buf = bun.default_allocator.alloc(u8, max_length) catch {
             globalObject.throwOutOfMemory();
             return .zero;
         };
 
-        if (prepend) |*pre| {
-            buf.appendSliceAssumeCapacity(pre.bytes[0..pre.len]);
-        }
+        const count = bun.simdutf.convert.utf16.to.utf8.le(remain, buf);
```

Comment:
I don't think we can use the non-validating version of this function

Reply:
Updated to use validating version